relay-server/src/actors/outcome_aggregator.rs

```diff
@@ -0,0 +1,167 @@
+
+use crate::statsd::RelayTimers;
+
+use super::outcome::{Outcome, OutcomeError, TrackOutcome};
+
+/// Contains everything to construct a `TrackOutcome`, except quantity
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct BucketKey {
+    /// Scoping of the request.
+    pub scoping: Scoping,
+    /// The outcome.
+    pub outcome: Outcome,
+    /// The event id.
+    pub event_id: Option<EventId>,
+    /// The client ip address.
+    pub remote_addr: Option<IpAddr>,
+    /// The event's data category.
+    pub category: DataCategory,
+}
+
+/// Aggregates outcomes into buckets, and flushes them periodically.
+/// Inspired by [`relay_metrics::Aggregator`].
+pub struct OutcomeAggregator {
+    /// The width of each aggregated bucket in seconds
+    bucket_interval: u64,
+    /// The time we should wait before flushing a bucket, in seconds
+    flush_delay: u64,
+    /// Maximum capacity of the actor's inbox
+    mailbox_size: usize,
+    /// Mapping from offset to bucket key to quantity. timestamp = offset * bucket_interval
+    buckets: BTreeMap<u64, HashMap<BucketKey, u32>>,
+    /// The recipient of the aggregated outcomes
+    outcome_producer: Recipient<TrackOutcome>,
+}
+
+impl OutcomeAggregator {
+    pub fn new(config: &Config, outcome_producer: Recipient<TrackOutcome>) -> Self {
+        // Set mailbox size to envelope buffer size, as in other global actors
+        let mailbox_size = config.envelope_buffer_size() as usize;
+
+        Self {
+            bucket_interval: config.outcome_bucket_interval(),
+            flush_delay: config.outcome_flush_delay(),
+            mailbox_size,
+            buckets: BTreeMap::new(),
+            outcome_producer,
+        }
+    }
+
+    fn do_flush(&mut self) {
+        let max_offset = (UnixTimestamp::now().as_secs() - self.flush_delay) / self.bucket_interval;
+        let bucket_interval = self.bucket_interval;
+        let outcome_producer = self.outcome_producer.clone();
+        self.buckets.retain(|offset, mapping| {
+            if offset <= &max_offset {
+                for (bucket_key, quantity) in mapping.drain() {
+                    let BucketKey {
+                        scoping,
+                        outcome,
+                        event_id,
+                        remote_addr,
+                        category,
+                    } = bucket_key;
+
+                    let timestamp = UnixTimestamp::from_secs(offset * bucket_interval);
+                    let outcome = TrackOutcome {
+                        timestamp: timestamp.as_datetime(),
+                        scoping,
+                        outcome,
+                        event_id,
+                        remote_addr,
+                        category,
+                        quantity,
+                    };
+
+                    relay_log::trace!("Flushing outcome for timestamp {}", timestamp);
+                    outcome_producer.do_send(outcome).ok(); // TODO: should we handle send errors here?
```

Comment:
Here we send messages in a loop. Could this overflow the outcome producer's mailbox? Should we send a batch of outcomes to the producer instead?

Reply:
Jan told me that this bypasses the inbox

Reply:
Right, thank you. The docs on `Recipient` are actually very clear on this:
https://docs.rs/actix/0.7.9/actix/struct.Recipient.html#method.do_send